### ISSUE TITLE
ACLs seem to not take affect correctly (IPv6, kubernetes)

### ISSUE THREAD

**Issue by nicoschottelius-lf** [coredns/coredns#4697]
**What happened**:

ACLs have been defined to *block* certain traffic as follows:

```
  Corefile: |
    .:53 {
        acl k8s.place7.ungleich.ch {
             allow net ::/0
        }
        acl . {
              allow net 2a0a:e5c0:13::/48
              block
        }
        forward . /etc/resolv.conf {
           max_concurrent 1000
        }
```

This configuration **does** allow the use of CoreDNS as an Open resolver from the whole IPv6 Internet.
It should not, though. I've tested this from the world wild web on different endpoints which are certainly out of this range and I was able to resolv any domain name.

**What you expected to happen**:

These ACLs should, in our opinion, do the following:

* allow any kind of request from 2a0a:e5c0:13::/48
* allow querying k8s.place7.ungleich.ch (the k8s domain) 
* block all other requests

**How to reproduce it (as minimally and precisely as possible)**:

Use above config 

**Anything else we need to know?**:

**Environment**:

- the version of CoreDNS: k8s.gcr.io/coredns/coredns:v1.8.0
- Corefile: See above
- logs, if applicable:
- OS (e.g: `cat /etc/os-release`):
- Others:

**Comment by telmich** [comment]
A bit of context of what we are trying to do, it might be helpful for understand the use case: https://ungleich.ch/u/blog/kubernetes-making-dns-publicly-reachable/

**Comment by ihac** [comment]
@nicoschottelius-lf @telmich Thanks for reporting the issue.

I tried similar config locally but failed to reproduce the issue you encountered:
```
.:53 {
    acl google.com {
        allow net 0.0.0.0/0
    }
    acl . {
        allow net 192.168.49.0/24
        block
    }
    forward . 114.114.114.114
}
```

```
.:53 {
    acl google.com {
        allow net ::/0
    }
    acl . {
        allow net fe80::4c:754c:46ee:e7d9/48
        block
    }
    forward . 114.114.114.114
}
```

**Comment by ihac** [comment]
But somehow I noticed there was a corner case where we could bypass the ACL check: if the source IP is illegal, ACL will never explicitly block the DNS query (hence it's allowed).

Source code: https://github.com/coredns/coredns/blob/master/plugin/acl/acl.go#L99. Here `ip` is nil when the `state.IP()` returns an illegal IP address.

I triggered this issue because I was sending DNS queries to my local CoreDNS server. So the src ip it actually get is `fe80::4c:754c:46ee:e7d9%en0`. The tailing `%en0` is used for routing link-local address but breaks the IP parsing logic (`net.ParseIP`: https://golang.org/pkg/net/#ParseIP)

**Comment by ihac** [comment]
I'd suggest you check once whether we encountered the same problem. Meanwhile, we'll fix the potential illegal src ip issue. cc @yongtang  @miekg

**Comment by chrisohaver** [comment]
https://en.wikipedia.org/wiki/IPv6_address#Scoped_literal_IPv6_addresses_(with_zone_index)
According to the link above, it is handled differently for Linux and BSD ... BSD encodes the info ~into the link local prefix~ in the address, resulting in a valid IP.

**Edit:** @ihac observes this is actually not true [below](https://github.com/coredns/coredns/issues/4697#issuecomment-862008096).

**Comment by telmich** [comment]
@ihac Would it help if I grant you access to the IPv6 only test cluster that I am using where I found the bug?

How I tested/triggered it, starting from a fresh IPv6 only cluster, standard Corefile:

```
[20:48] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system get svc
NAME       TYPE        CLUSTER-IP           EXTERNAL-IP   PORT(S)                  AGE
kube-dns   ClusterIP   2a0a:e5c0:13:e2::a   <none>        53/UDP,53/TCP,9153/TCP   24h
[20:49] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system describe cm coredns
Name:         coredns
Namespace:    kube-system
Labels:       <none>
Annotations:  <none>

Data
====
Corefile:
----
.:53 {
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes c2.k8s.ooo in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf {
       max_concurrent 1000
    }
    cache 30
    loop
    reload
    loadbalance
}

Events:  <none>

```

So far, nothing should stop anyone from querying. Editing the config & deleting the original pods:

```
[20:49] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system edit cm coredns
configmap/coredns edited
[20:51] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system describe cm coredns
Name:         coredns
Namespace:    kube-system
Labels:       <none>
Annotations:  <none>

Data
====
Corefile:
----
.:53 {
    acl c2.k8s.ooo {
         allow net ::/0
    }
    acl . {
          allow net 2a0a:e5c0:13::/48
          block
    }
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes c2.k8s.ooo in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf {
       max_concurrent 1000
    }
    cache 30
    loop
    reload
    loadbalance
}

Events:  <none>
[20:52] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system get pods -l k8s-app=kube-dns
NAME                       READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-56dv9   1/1     Running   1          25h
coredns-558bd4d5db-hsspb   1/1     Running   1          25h
[20:52] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system delete pods -l k8s-app=kube-d
ns
pod "coredns-558bd4d5db-56dv9" deleted
pod "coredns-558bd4d5db-hsspb" deleted
[20:52] server47.place7:~/ungleich-k8s/rook# kubectl -n kube-system get pods -l k8s-app=kube-dns
NAME                       READY   STATUS    RESTARTS   AGE
coredns-558bd4d5db-74fzl   1/1     Running   0          38s
coredns-558bd4d5db-btpfp   1/1     Running   0          38s
[20:53] server47.place7:~/ungleich-k8s/rook# 

```

Testing from  a remote VM:

```
root@nico-vm:~# dig ungleich.ch @2a0a:e5c0:13:e2::a

; <<>> DiG 9.11.5-P4-5.1+deb10u1-Debian <<>> ungleich.ch @2a0a:e5c0:13:e2::a
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 10308
;; flags: qr aa rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 14cabb8afab2a48a (echoed)
;; QUESTION SECTION:
;ungleich.ch.			IN	A

;; ANSWER SECTION:
ungleich.ch.		30	IN	A	185.203.112.17

;; Query time: 19 msec
;; SERVER: 2a0a:e5c0:13:e2::a#53(2a0a:e5c0:13:e2::a)
;; WHEN: Tue Jun 15 20:55:46 CEST 2021
;; MSG SIZE  rcvd: 79

root@nico-vm:~# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 9000 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 02:00:b3:39:7b:eb brd ff:ff:ff:ff:ff:ff
    inet6 2a0a:e5c0:0:2:400:b3ff:fe39:7beb/64 scope global 
       valid_lft forever preferred_lft forever
    inet6 fe80::b3ff:fe39:7beb/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc pfifo_fast state UP group default qlen 1000
    link/ether 02:00:b9:cb:70:0a brd ff:ff:ff:ff:ff:ff
    inet 185.203.112.10/24 brd 185.203.112.255 scope global eth1
       valid_lft forever preferred_lft forever
    inet6 fe80::b9ff:fecb:700a/64 scope link 
       valid_lft forever preferred_lft forever
root@nico-vm:~# 
```

**Comment by ihac** [comment]
> Of note, it is handled differently for Linux and BSD ... BSD encodes the info into the link local prefix in the address, resulting in a valid IP.

@chrisohaver Hi I'm not quite sure about this, but I tested it in both macOS & Linux, and the link-local address was still invalid:
```
$ dig -6 @fe80::4001:aff:feaa:4 www.google.com

state.IP() returns `fe80::4001:aff:feaa:4%ens4`.
```

**Comment by ihac** [comment]
@telmich
> Would it help if I grant you access to the IPv6 only test cluster that I am using where I found the bug?

Yeh it helps. Also, let's narrow down the possibilities by adding a logging statement (sth like `fmt.Println(state.IP(), ip)`) in CoreDNS https://github.com/coredns/coredns/blob/master/plugin/acl/acl.go#L100. Can you try a few more queries and post the logs here?
This helps us to understand whether the problem emerges from the ACL setup process or the match process.

**Comment by telmich** [comment]
@ihac I pinged you on slack, but also here for reference: if you can give me your public ssh key, I can grant you access to it

**Comment by ihac** [comment]
@telmich Not sure where and which account you pinged. I'm in #coredns with name `Luffy Xiao`.

**Comment by telmich** [comment]
Currently testing with a new cluster on p10.k8s.ooo.

Corefile:

```
.:53 {
     acl p10.k8s.ooo {
         allow net ::/0
     }
     acl . {
       allow net 2a0a:e5c0:10:2::/64
       block
    }
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes p10.k8s.ooo in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf {
       max_concurrent 1000
    }
    cache 30
    loop
    reload
    loadbalance
}
```

Getting the DNS service IP:

```
[18:53] nb3:~% kubectl -n kube-system get svc            
NAME       TYPE        CLUSTER-IP          EXTERNAL-IP   PORT(S)                  AGE
kube-dns   ClusterIP   2a0a:e5c0:10:3::a   <none>        53/UDP,53/TCP,9153/TCP   67m
```

Checking from an outside IP address:

```
[18:53:46] nico-vm:~$ dig ungleich.ch @2a0a:e5c0:10:3::a aaaa

; <<>> DiG 9.11.5-P4-5.1+deb10u5-Debian <<>> ungleich.ch @2a0a:e5c0:10:3::a aaaa
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: REFUSED, id: 54598
;; flags: qr rd; QUERY: 1, ANSWER: 0, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 212435e6d019d30b (echoed)
;; QUESTION SECTION:
;ungleich.ch.                   IN      AAAA

;; Query time: 1 msec
;; SERVER: 2a0a:e5c0:10:3::a#53(2a0a:e5c0:10:3::a)
;; WHEN: Sat Aug 14 18:53:48 CEST 2021
;; MSG SIZE  rcvd: 52

[18:53:48] nico-vm:~$ dig kube-dns.kube-system.svc.p10.k8s.ooo @2a0a:e5c0:10:3::a aaaa

; <<>> DiG 9.11.5-P4-5.1+deb10u5-Debian <<>> kube-dns.kube-system.svc.p10.k8s.ooo @2a0a:e5c0:10:3::a aaaa
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 58921
;; flags: qr aa rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1
;; WARNING: recursion requested but not available

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: dfb2e8bad33f638c (echoed)
;; QUESTION SECTION:
;kube-dns.kube-system.svc.p10.k8s.ooo. IN AAAA

;; ANSWER SECTION:
kube-dns.kube-system.svc.p10.k8s.ooo. 30 IN AAAA 2a0a:e5c0:10:3::a

;; Query time: 1 msec
;; SERVER: 2a0a:e5c0:10:3::a#53(2a0a:e5c0:10:3::a)
;; WHEN: Sat Aug 14 18:54:07 CEST 2021
;; MSG SIZE  rcvd: 141
```

This looks like correct behaviour!

I am running the image `k8s.gcr.io/coredns/coredns:v1.8.4`

**Comment by telmich** [comment]
@ihac @chrisohaver Can we reopen this bug? It seems 1.8.4 is working correctly, but on another cluster with 1.8.6 the problem is happening again.

The following CM still makes coredns behave as an open resolver:

```
 kubectl -n kube-system describe cm coredns                  
Name:         coredns
Namespace:    kube-system
Labels:       <none>
Annotations:  <none>

Data
====
Corefile:
----
.:53 {
    acl k8s01.via.de {
      allow net ::/0
    }
    acl . {
      allow net 2a01:9820::/32
      block
    }
    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes k8s01.via.de in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf {
       max_concurrent 1000
    }
    cache 30
    loop
    reload
    loadbalance
}

```

Querying coredns for a domain that it is not responsible for and it certainly hasn't cached either:

```
[10:45] blind:~% dig matrix.ungleich.ch aaaa @kube-dns.kube-system.svc....          

; <<>> DiG 9.16.27 <<>> matrix.ungleich.ch aaaa @kube-dns.kube-system.svc....
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 33289
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 4096
; COOKIE: 8fa9614bdcf424cc (echoed)
;; QUESTION SECTION:
;matrix.ungleich.ch.		IN	AAAA

;; ANSWER SECTION:
matrix.ungleich.ch.	30	IN	AAAA	2a0a:e5c0:2:12:0:f0ff:fea9:c45f

;; Query time: 46 msec
;; SERVER: 2a01:9820:6:11::a#53(2a01:9820:6:11::a)
;; WHEN: Mon Apr 04 10:45:17 CEST 2022
;; MSG SIZE  rcvd: 105
```

**Comment by telmich** [comment]
@chrisohaver Do you know if there was a change to the codepath between 1.8.4 and 1.8.6? I can verify that 2 1.8.4 clusters work correctly, while two 1.8.6 clusters are acting as open resolvers.

As all clusters are IPv6 reachable, this is somewhat problematic. We cannot filter incoming DNS, as the clusterDomain is used for accessing services from outside: customers are accessing services via `servicename.namespace.svc.XYZ.k8s.ooo`, thus requiring access to coredns.

**Comment by chrisohaver** [comment]
> @chrisohaver Do you know if there was a change to the codepath between 1.8.4 and 1.8.6?

I don’t, but I can look at the commits between the two releases.

**Comment by chrisohaver** [comment]
> @chrisohaver Do you know if there was a change to the codepath between 1.8.4 and 1.8.6?

I didn't find any changes directly to the _acl_ plugin in 1.8.5 or 1.8.6.

@telmich, if you drop the clusters running 1.8.6 down to 1.8.4, do they start working as expected?

@ihac, do you recall which version you tested [when you tried (can could not) replicate the issue](https://github.com/coredns/coredns/issues/4697#issuecomment-861487718)?

**Comment by chrisohaver** [comment]
Reviewing the history of the comments above ...

* problem observed in 1.8.0 in a k8s cluster (cluster A)
* problem no longer observed in 1.8.4, in a "new cluster" (cluster B)
* problem observed again in 1.8.6, in a "different cluster" (cluster C)

It would be good to verify that in a single given cluster we can show that one version of CoreDNS works, and another does not.

**Comment by telmich** [comment]
TL;DR: p5 w/ 1.8.4 also presents itself as an open resolver.

So I just found an existing cluster running 1.8.4, with the following CM:

```
[21:07] nb3:~% kubectl -n kube-system describe cm coredns
Name:         coredns
Namespace:    kube-system
Labels:       <none>
Annotations:  <none>

Data
====
Corefile:
----
.:53 {
    acl p5.k8s.ooo {
       allow net ::/0
    }
    acl . {
       allow net 2a0a:e5c0::/29
       block
    }

    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes p5.k8s.ooo in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf {
       max_concurrent 1000
    }
    cache 30
    loop
    reload
    loadbalance
}


BinaryData
====

Events:  <none>
```

Querying it from the outside

```
[21:11] emonitor1.place12:~# dig +short ungleich.ch @kube-dns.kube-system.svc.p5.k8s.ooo
185.203.112.17
[21:12] emonitor1.place12:~# dig +short google.com @kube-dns.kube-system.svc.p5.k8s.ooo
142.250.203.110
[21:12] emonitor1.place12:~# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
    link/ether 00:0d:b9:5a:7f:b0 brd ff:ff:ff:ff:ff:ff
    inet 192.168.178.43/24 brd 192.168.178.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2003:a:311:8500:20d:b9ff:fe5a:7fb0/64 scope global dynamic flags 100 
       valid_lft 7058sec preferred_lft 1079sec
    inet6 fe80::20d:b9ff:fe5a:7fb0/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether 00:0d:b9:5a:7f:b1 brd ff:ff:ff:ff:ff:ff
4: eth2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether 00:0d:b9:5a:7f:b2 brd ff:ff:ff:ff:ff:ff
[21:13] emonitor1.place12:~# 

```

This seems to indicate that we did incorrectly conclude that 1.8.4 actually did introduce a fix. At least it seems that in this is the case in that cluster. Or is it that the ACL parsing code does not handle /29's or /32's correctly?

I'll check again with another cluster in which I was pretty sure that 1.8.4 was working correctly.

**Comment by telmich** [comment]
TL;DR: p10 w/ 1.8.4 also presents itself as an open resolver.

```
[21:17] nb3:~% kubectl -n kube-system describe cm coredns              
Name:         coredns
Namespace:    kube-system
Labels:       <none>
Annotations:  <none>

Data
====
Corefile:
----
.:53 {
    acl p10.k8s.ooo {
        allow net ::/0
    }
    acl . {
        allow net 2a0a:e5c0::/29
        block
    }

    errors
    health {
       lameduck 5s
    }
    ready
    kubernetes p10.k8s.ooo in-addr.arpa ip6.arpa {
       pods insecure
       fallthrough in-addr.arpa ip6.arpa
       ttl 30
    }
    prometheus :9153
    forward . /etc/resolv.conf {
       max_concurrent 1000
    }
    cache 30
    loop
    reload
    loadbalance
    dns64 {
      prefix 2a0a:e5c0:2:10::/96
    }
}


BinaryData
====

Events:  <none>
[21:20] nb3:~% kubectl describe deployments.apps -n kube-system coredns | grep Image:
    Image:       k8s.gcr.io/coredns/coredns:v1.8.4
```

Querying from the outside:

```
[21:20] emonitor1.place12:~# dig +short google.com @kube-dns.kube-system.svc.p10.k8s.ooo
172.217.168.46
[21:20] emonitor1.place12:~# dig +short ungleich.ch @kube-dns.kube-system.svc.p10.k8s.ooo
185.203.112.17
[21:20] emonitor1.place12:~# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2: eth0: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc mq state UP qlen 1000
    link/ether 00:0d:b9:5a:7f:b0 brd ff:ff:ff:ff:ff:ff
    inet 192.168.178.43/24 brd 192.168.178.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 2003:a:311:8500:20d:b9ff:fe5a:7fb0/64 scope global dynamic flags 100 
       valid_lft 6851sec preferred_lft 1451sec
    inet6 fe80::20d:b9ff:fe5a:7fb0/64 scope link 
       valid_lft forever preferred_lft forever
3: eth1: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether 00:0d:b9:5a:7f:b1 brd ff:ff:ff:ff:ff:ff
4: eth2: <NO-CARRIER,BROADCAST,MULTICAST,UP> mtu 1500 qdisc mq state DOWN qlen 1000
    link/ether 00:0d:b9:5a:7f:b2 brd ff:ff:ff:ff:ff:ff
[21:22] emonitor1.place12:~# 
```

**Comment by telmich** [comment]
So in a summary, open resolvers with the described configurations can be found in:

* 1.8.0
* 1.8.4
* 1.8.6

hey, at least all tested ones are even...

**Comment by telmich** [comment]
I also checked this morning that a /48 netmask does not fix the issue either (I suspect incorrect netmask calculations)...

...wait, something just comes to my mind: the ACLs might actually work, just not inside kubernetes. In all clusters we are using calico as the CNI with BGP peering with upstream. How calico works by default is that *every node* announces the service IP range.

The kubernetes cluster / calico itself is then forwarding to the correct node on which a correct pod is running. This does not keep the source address intact and thus coredns *might/will* see intra cluster source IP addresses.

I've to verify the hypothesis, but this would be a very likely source of the problem.

Is there a way to log/print every request in coredns so that we see the source IP address?

**Comment by chrisohaver** [comment]
> I've to verify the hypothesis, but this would be a very likely source of the problem.

Yes, now that you conclude it never actually worked in 1.8.4 (or any other version you tried).

> Is there a way to log/print every request in coredns so that we see the source IP address?

Yes, the _log_ plugin displays remote ip (source ip) as the first field (this actually logs _responses_ - so it would not work for logging dropped requests).

**Comment by ruben-herold** [comment]
I have done a little test:
I take a look on which worker a coredns pod runs. Then I connect to it as root and did an tcpdump on the calico interface.
And it shows that the query has has source ip the ip from the worker node...
So all ACL's are useless ...

**Comment by ihac** [comment]
Sorry I missed the GitHub notifications.

> @ihac, do you recall which version you tested https://github.com/coredns/coredns/issues/4697#issuecomment-861487718?

@chrisohaver I believe it's v1.8.0, for which @telmich reported the issue.

And from what I observed after debugging the issue locally, an invalid source IP address could simply bypass all ACL rules. So CoreDNS will either block or accept all the incoming requests, based on the default rule configured in ACL.

> wait, something just comes to my mind: the ACLs might actually work, just not inside kubernetes. In all clusters we are using calico as the CNI with BGP peering with upstream. How calico works by default is that every node announces the service IP range.

This seems to be a valid hypothesis. To verify this, I typically add a logging statement in the ACL plugin (dump the source IP) and re-build CoreDNS. Not quite sure whether this is feasible for you or not.

> I take a look on which worker a coredns pod runs. Then I connect to it as root and did an tcpdump on the calico interface.
And it shows that the query has has source ip the ip from the worker node...
> So all ACL's are useless ...

Sadly true. if the source IP extracted from the incoming DNS queries in CoreDNS is not the true client IP, ACL will never work as we expect.

**Comment by ruben-herold** [comment]
So I think we can close it here and make some notice about the behaviour in the documentation. 
I have now create a workaround:

- Install seperate coredns app on the cluster with seperate service 
- forward the zones to the kube-dns service
- forbid recursion via ACL for all source
- delegate the domains to this service
- make the kube-dns not reachable from outside.

Seems to be the only solution as long there is no possibility to keep the source ip.

**Comment by yongtang** [comment]
Added a note in PR #5311.

**Comment by chrisohaver** [comment]
@telmich, Since the source IP is mutated before it reaches CoreDNS, the ACL (as described in [your blog](https://ungleich.ch/u/blog/kubernetes-making-dns-publicly-reachable/)) should be implemented in your BIND forwarder instead: https://bind9.readthedocs.io/en/v9_16_4/security.html

**Comment by ruben-herold** [comment]
@chrisohaver  in the setup described there is no BIND forwarder. The is a direct delegation for a domain to the kube-dns.

**Comment by chrisohaver** [comment]
> @chrisohaver in the setup described there is no BIND forwarder. The is a direct delegation for a domain to the kube-dns.

OK.  I assumed the following config snip was served by a BIND server.

```
zone "k8s.place7.ungleich.ch"  {
   type forward;
   forward only;
   forwarders { 2a0a:e5c0:13:e2::a; };
};
```

**Comment by telmich** [comment]
@chrisohaver That was one of the many tries to make it working - atm, there are quite some coredns instances publicly reachable. @ruben-herold already checked with the calico guys and as a first step documenting this "non feature" on coredns side is probably a good approach.

Short/mid term we should probably see if Calico cannot be patched to preserve the IP without eBPF *or* to switch to eBPF with IPv6 support.
